### PR TITLE
New Feature - Dialog Maker

### PR DESCRIPTION
**Pull request by not-archiey**
This Feature will allow users to create a dialog image (camp buddy style) with their own preferences of characters, backgrounds, and text.

Generated Image Example

### CODE REVIEW

File: commands/Dialog.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import Discord from "discord.js";
+import {
+  EMOJI_REGEX,
+  EMOTE_MENTIONS_REGEX,
+  NONASCII_REGEX
+} from "./base/DiscordRegex";
+import Command from "./base/Command";
+
+import backgrounds from "../assets/images/dialog/bg/backgrounds.json";
+import characters from "../assets/images/dialog/char/characters.json";
```
Comment (duncte123) [inline commands/Dialog.js]: You could move these one folder up
```suggestion
import backgrounds from "../assets/images/dialog/backgrounds.json";
import characters from "../assets/images/dialog/characters.json";
```

File: commands/Dialog.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      category: "fun",
+      description: "Gives you an image of a dialog"
+    });
+  }
+
+  async run(_client, message, args) {
+    const now = Date.now();
+
+    if (args.length < 2) {
+      message.channel.send("Sorry, but i didn't understand that~!");
```
Comment (duncte123) [inline commands/Dialog.js]: Users will probably not understand what you mean by this

File: commands/Dialog.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      return;
+    }
+
+    // Gets the character argument by splitting the first args
+    const arg0 = args[0].toLowerCase();
+    const arg1 = args[1].toLowerCase();
```
Comment (duncte123) [inline commands/Dialog.js]: Be more descriptive with these variable names please

File: commands/Dialog.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    // Where the message starts from the arglist, defaults to 1
+    let messageStart = 1;
+
+    let img;
+
+    // Checks if arg0 is a background and arg1 is a character
+    if (backgrounds[arg0] !== undefined && characters[arg1] !== undefined) {
```
Comment (duncte123) [inline commands/Dialog.js]: Please do null checks instead

File: commands/Dialog.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+          `Sorry, I have ${arg0} for the background but I dont have that character!`
+        );
+      } else {
+        // If neither is found
+        message.channel.send(
+          `Sorry, I dont have that character or background~!`
+        );
+      }
+      return;
+    }
```
Comment (duncte123) [inline commands/Dialog.js]: This entire if block looks too complicated
try doing the image cheks before you do anything else

File: commands/Dialog.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    // Tests if message includes emoji or emotes
+    if (
+      EMOJI_REGEX.test(messageToSay) ||
+      EMOTE_MENTIONS_REGEX.test(messageToSay) ||
+      NONASCII_REGEX.test(messageToSay)
+    ) {
+      message.channel.send(
+        "Sorry, this bot cannot display emotes, mentions or non-latin characters"
```
Comment (duncte123) [inline commands/Dialog.js]: Let's keep it a bit fun, talk from the bot's perspective instead

File: commands/Dialog.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      );
+      return;
+    }
+
+    // Selects a bitmap font based on the message length
+    const font = await Jimp.loadFont("assets/fonts/halogen_56.fnt");
```
Comment (duncte123) [inline commands/Dialog.js]: Why do you have many font files included when you only use one?

**Comment by duncte123** [comment]
The images generated are 810*1080, they don't need to be that large if you ask me

File: commands/Dialog.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      // If firstArg is not a correct bg but right chara
+      message.channel.send(
+        `Sorry, I do have ${firstArg} but I don't have ${secondArg}~!`
+      );
+      return;
+    } else {
+      // If both is unrecognizable | Since parsing secondArg could be a message, text is like this
+      message.channel.send(`Sorry, I dont have that character or background~!`);
+      return;
+    }
```
Comment (duncte123) [inline commands/Dialog.js]: There is no need for else branches if you keep returning, I also suggest moving the character and location to variables as you did with message start so you only have to do `Jimp.read` once

This entire if-statement could probably just be
```js
const background = backgrounds[firstArg] || 'camp';
const character = characters[firstArg] || characters[secondArg];

if (!background) {
  message.channel.send('Whoops sorry I don\'t have that background');
  retrurn;
}

if (!character) {
  message.channel.send('I don\'t think that the character you want is in Camp Buddy');
  retrurn;
}

const img = await Jimp.read(`./cache/${character}_${background}.png`);```

File: commands/Dialog.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    // Gets the character argument by splitting the first args
+    const firstArg = args[0].toLowerCase();
+    const secondArg = args[1].toLowerCase();
+    // Where the message starts from the arglist, defaults to 1
+    let messageStart = 1;
+
+    let img;
+
+    if (characters[firstArg] != null) {
```
Comment (duncte123) [inline commands/Dialog.js]:
```suggestion
    if (characters[firstArg]) {
```

File: commands/Dialog.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+      // If firstArg is a character
+      img = await Jimp.read(`./cache/${firstArg}_camp.png`);
+    } else if (backgrounds[firstArg] != null && characters[secondArg] != null) {
+      // If firstArg is a background and secondArg is a character
+      img = await Jimp.read(`./cache/${secondArg}_${firstArg}.png`);
+      messageStart = 2;
+    } else if (backgrounds[firstArg] == null && characters[secondArg] != null) {
```
Comment (duncte123) [inline commands/Dialog.js]:
```suggestion
    } else if (!backgrounds[firstArg] && characters[secondArg]) {
```

File: commands/Dialog.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    const secondArg = args[1].toLowerCase();
+    // Where the message starts from the arglist, defaults to 1
+    let messageStart = 1;
+
+    let img;
+
+    if (characters[firstArg] != null) {
+      // If firstArg is a character
+      img = await Jimp.read(`./cache/${firstArg}_camp.png`);
+    } else if (backgrounds[firstArg] != null && characters[secondArg] != null) {
```
Comment (duncte123) [inline commands/Dialog.js]:
```suggestion
    } else if (backgrounds[firstArg] && characters[secondArg]) {
```

File: utils/discordPatterns.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+// A regex string to ban emotes and emojis from this command
+export const EMOJI_REGEX = new RegExp(
+  "(?:[\u2700-\u27bf]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff]|[\u0023-\u0039]\ufe0f?\u20e3|\u3299|\u3297|\u303d|\u3030|\u24c2|\ud83c[\udd70-\udd71]|\ud83c[\udd7e-\udd7f]|\ud83c\udd8e|\ud83c[\udd91-\udd9a]|\ud83c[\udde6-\uddff]|\ud83c[\ude01-\ude02]|\ud83c\ude1a|\ud83c\ude2f|\ud83c[\ude32-\ude3a]|\ud83c[\ude50-\ude51]|\u203c|\u2049|[\u25aa-\u25ab]|\u25b6|\u25c0|[\u25fb-\u25fe]|\u00a9|\u00ae|\u2122|\u2139|\ud83c\udc04|[\u2600-\u26FF]|\u2b05|\u2b06|\u2b07|\u2b1b|\u2b1c|\u2b50|\u2b55|\u231a|\u231b|\u2328|\u23cf|[\u23e9-\u23f3]|[\u23f8-\u23fa]|\ud83c\udccf|\u2934|\u2935|[\u2190-\u21ff])",
+  "g"
+);
+export const EMOTE_MENTIONS_REGEX = new RegExp(/<(?:[^\d>]+|:[A-Za-z0-9]+:)\w+>/);
+export const NONASCII_REGEX = new RegExp(/[^\x00-\x7F]/);
```
Comment (duncte123) [inline utils/discordPatterns.js]: Please put a new line at the end of the file

File: package.json [hunk 13]
```diff
@@ -11,6 +11,8 @@
     "lint:fix": "eslint . --fix"
   },
   "dependencies": {
+    "esm": "^3.2.25",
```
Comment (duncte123) [inline package.json:14]: We don't use esm anymore, make sure to merge develop into your branch

Comment (not-archiey) [reply package.json:14]: Whoops sorry, I've missed that while resolving the git conflicts

File: commands/Dialog.js [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+      return;
+    }
+
+    const body = {
+      background,
+      character: charArg,
+      text: messageToSay
+    };
+
+    console.log(body);
```
Comment (duncte123) [inline commands/Dialog.js:136]: This should be removed

File: utils/discordPatterns.js [hunk 15]
```diff
@@ -1,2 +1,12 @@
+// A regex string to ban emotes and emojis from this command
+export const EMOJI_REGEX = new RegExp(
+  "(?:[\u2700-\u27bf]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff]|[\u0023-\u0039]\ufe0f?\u20e3|\u3299|\u3297|\u303d|\u3030|\u24c2|\ud83c[\udd70-\udd71]|\ud83c[\udd7e-\udd7f]|\ud83c\udd8e|\ud83c[\udd91-\udd9a]|\ud83c[\udde6-\uddff]|\ud83c[\ude01-\ude02]|\ud83c\ude1a|\ud83c\ude2f|\ud83c[\ude32-\ude3a]|\ud83c[\ude50-\ude51]|\u203c|\u2049|[\u25aa-\u25ab]|\u25b6|\u25c0|[\u25fb-\u25fe]|\u00a9|\u00ae|\u2122|\u2139|\ud83c\udc04|[\u2600-\u26FF]|\u2b05|\u2b06|\u2b07|\u2b1b|\u2b1c|\u2b50|\u2b55|\u231a|\u231b|\u2328|\u23cf|[\u23e9-\u23f3]|[\u23f8-\u23fa]|\ud83c\udccf|\u2934|\u2935|[\u2190-\u21ff])",
+  "g"
+);
+export const EMOTE_MENTIONS_REGEX = new RegExp(
+  /<(?:[^\d>]+|:[A-Za-z0-9]+:)\w+>/
+);
+// eslint-disable-next-line no-control-regex
+export const NONASCII_REGEX = new RegExp(/[^\x00-\x7F]/);
```
Comment (duncte123) [inline utils/discordPatterns.js:10]:
```suggestion
export const NONASCII_REGEX = /[^\x00-\x7F]/;
```

Comment (duncte123) [reply utils/discordPatterns.js:10]: same for the thing above

File: utils/discordPatterns.js [hunk 16]
```diff
@@ -1,2 +1,7 @@
+// A regex string to ban emotes and emojis from this command
+export const EMOJI_REGEX = /(?:[\u2700-\u27bf]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff]|[\u0023-\u0039]\ufe0f?\u20e3|\u3299|\u3297|\u303d|\u3030|\u24c2|\ud83c[\udd70-\udd71]|\ud83c[\udd7e-\udd7f]|\ud83c\udd8e|\ud83c[\udd91-\udd9a]|\ud83c[\udde6-\uddff]|\ud83c[\ude01-\ude02]|\ud83c\ude1a|\ud83c\ude2f|\ud83c[\ude32-\ude3a]|\ud83c[\ude50-\ude51]|\u203c|\u2049|[\u25aa-\u25ab]|\u25b6|\u25c0|[\u25fb-\u25fe]|\u00a9|\u00ae|\u2122|\u2139|\ud83c\udc04|[\u2600-\u26FF]|\u2b05|\u2b06|\u2b07|\u2b1b|\u2b1c|\u2b50|\u2b55|\u231a|\u231b|\u2328|\u23cf|[\u23e9-\u23f3]|[\u23f8-\u23fa]|\ud83c\udccf|\u2934|\u2935|[\u2190-\u21ff])/g;
+export const EMOTE_MENTIONS_REGEX = /<(?:[^\d>]+|:[A-Za-z0-9]+:)\w+>/;
+// eslint-disable-next-line no-control-regex
```
Comment (duncte123) [inline utils/discordPatterns.js:4]: is this still needed?

Comment (not-archiey) [reply utils/discordPatterns.js:4]: Extra redundancy i guess, but if you want it to be removed I'll remove it...

File: commands/Dialog.js [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+        "Sorry, I can't display emotes, mentions, or non-latin characters"
+      );
+      return;
+    }
+
+    const body = {
+      background,
+      character,
+      text
+    };
```
Comment (duncte123) [inline commands/Dialog.js:130]: Is this allowed to be inlined?

**Pull request merged**